lib/limiter/internal/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,193 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package ratelimit
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// TokenLimiter implements rate limiting middleware.
+type TokenLimiter struct {
+	clock clockwork.Clock
+	log   *slog.Logger
+
+	defaultRates *RateSet
+
+	mutex      sync.Mutex
+	bucketSets *utils.FnCache
+	next       http.Handler
+}
+
+type TokenLimiterConfig struct {
+	Rates *RateSet
+	Clock clockwork.Clock
+}
+
+// New constructs a `TokenLimiter` middleware instance.
+func New(config TokenLimiterConfig) (*TokenLimiter, error) {
+	if len(config.Rates.m) == 0 {
+		return nil, trace.BadParameter("Missing required rates")
+	}
+	clock := config.Clock
+	if clock == nil {
+		clock = clockwork.NewRealClock()
+	}
+	tl := &TokenLimiter{
+		defaultRates: config.Rates,
+		clock:        clock,
+
+		log: slog.With(teleport.ComponentKey, "ratelimiter"),
+	}
+
+	bucketSets, err := utils.NewFnCache(utils.FnCacheConfig{
+		// The default TTL here is not super important because we set the
+		// TTL explicitly for each entry we insert.
+		TTL:   10 * time.Second,
+		Clock: clock,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tl.bucketSets = bucketSets
+	return tl, nil
+}
+
+func (tl *TokenLimiter) Wrap(next http.Handler) {
+	tl.next = next
+}
+
+func (tl *TokenLimiter) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	clientIP, err := ExtractClientIP(req)
+	if err != nil {
+		ServeHTTPError(w, req, err)
+		return
+	}
+
+	if err := tl.consumeRates(clientIP, 1 /* amount */); err != nil {
+		tl.log.InfoContext(context.Background(), "limiting request",
+			"method", req.Method, "url", req.URL.String(), "error", err)
+		ServeHTTPError(w, req, err)
+		return
+	}
+
+	tl.next.ServeHTTP(w, req)
+}
+
+func (tl *TokenLimiter) consumeRates(source string, amount int64) error {
+	tl.mutex.Lock()
+	defer tl.mutex.Unlock()
+
+	effectiveRates := tl.defaultRates
+
+	// We set the TTL as 10 times the rate period. E.g. if rate is 100 requests/second
+	// per client IP, the counters for this IP will expire after 10 seconds of inactivity.
+	ttl := effectiveRates.MaxPeriod()*10 + 1
+	bucketSet, err := utils.FnCacheGetWithTTL(context.TODO(), tl.bucketSets, source, ttl,
+		func(ctx context.Context) (*TokenBucketSet, error) {
+			return NewTokenBucketSet(effectiveRates, tl.clock), nil
+		},
+	)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	bucketSet.Update(effectiveRates)
```

Comment:
Note: in the original info, this used to only happen if the bucket set was already found in the cache.

With `FnCache` it now happens in both cases. I don't think this should be a problem, but appreciate extra eyes when reviewing this part.